# Laboratorio 2 - Parte 2

### Modelos no paramétricos

### 2018-II

#### Profesor: Julián D. Arias Londoño
#### julian.ariasl@udea.edu.co


## Guía del laboratorio

En esta archivo va a encontrar tanto celdas de código cómo celdas de texto con las instrucciones para desarrollar el laboratorio.

Lea atentamente las instrucciones entregadas en las celdas de texto correspondientes y proceda con la solución de las preguntas planteadas.

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

In [2]:
#Tomado de https://github.com/rramosp/mooc-grader
from Autentication import *#python 3
import inspect, urllib
html, auth_code, userinfo = google_authenticate(PORT_NUMBER=8080)
html

Haga click en el siguiente enlace para autenticarse con su cuenta de correo institucional
https://accounts.google.com/o/oauth2/auth?scope=profile+email&client_id=893762525034-g9d91ddls9e19a1q77c7hsq2rhgqo9h7.apps.googleusercontent.com&access_type=offline&redirect_uri=http%3A%2F%2Flocalhost%3A8080&response_type=code
waiting for authentication ...
authentication succeeded
/?code=4/TgDA9tXMcGU1db2RtY5ZmPyRPl8zOG6eG6AZXx_dTVzVE4ZHSrs_8jskpfyp9C2Is4jTneVOIgMt0tWbRccmm2c


#### Segundo integrante:

David Alejandro Marín Alzate

In [2]:
!pip install pixiedust

In [16]:
%matplotlib inline
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
from __future__ import division
from scipy import stats


#Algunas advertencias que queremos evitar
import warnings
warnings.filterwarnings("always")

## Ejercicio 1

Para el problema de clasificación usaremos la base de datos iris, que cargaremos desde la librería sklearn. Responda las siguientes preguntas:

In [5]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets

iris = datasets.load_iris()

#Esta es la base de datos Iris del UCI Machine Learning Repository 
#https://archive.ics.uci.edu/ml/datasets/iris

X, Y = iris.data, iris.target


/Users/alejandro/py3/lib/python3.7/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/Users/alejandro/py3/lib/python3.7/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/Users/alejandro/py3/lib/python3.7/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/Users/alejandro/py3/lib/python3.7/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/Users/alejandro/py3/lib/python3.7/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/Users/alejandro/py3/lib/

In [12]:
print("Muestras de entrada ", X.shape)
print("Diferente clases que hay en la salida ",np.unique(Y))
print("Muestras x clase ", np.unique(Y, return_counts=True))

Muestras de entrada  (150, 4)
Diferente clases que hay en la salida  [0 1 2]
Muestras x clase  (array([0, 1, 2]), array([50, 50, 50]))


1.1 ¿Cu&aacute;ntas clases tiene el problema?: 

La base de datos cuenta con **3 clases**, en este caso correponde a [0,1,2]

1.2 ¿Cu&aacute;ntas caracter&iacute;sticas tiene el problema?:

La base de datos de Iris cuenta con **4 carácteristicas**.

1.3 ¿Cu&aacute;ntas muestras tiene el problema?:

El problema presenta con **150 muestras**.

1.4 ¿Cu&aacute;ntas muestras por cada clase hay en la base de datos?:

Por cada clase hay **50 muestras**, se observa que la distribuición es equitativa.

<b>Nota</b>: Agregue una celda con el código usado para responder cada una de las preguntas anteriores.

## Ejercicio 2

Analice los siguientes métodos a la luz de la teoría vista para los modelos de K vecinos más cercanos (KNN) y complete el código del método KNN.

<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función scipy.spatial.distance.euclidean o usar la función numpy.linalg.norm. <b>Revisen la documentación </b> para comprender su uso. También serán de utilidad las funciones numpy.sort y numpy.argsort.

In [17]:
from scipy.stats import mode

import operator

#Para calcular el error en los problemas de clasificación
def ErrorClas(Y_lest, Y):
    error = 1 - np.sum(Y_lest == Y)/len(Y)
    
    return error


def KNN(X_train, Y_train, X_val, k, tipo):
    
    #X_train es la matriz con las muestras de entrenamiento
    #Y_train es un vector con los valores de salida pra cada una de las muestras de entrenamiento
    #X_val es la matriz con las muestras de validación
    #tipo es una bandera que indica si el problema es de regresión o de clasificación.
    tipo = tipo
    k = k    #Parámetro k que equivale al número de vecinos a tener en cuenta para resolver el problema de
    #predicción de la variable de salida
    
    N = np.shape(Y_train)[0]
    Y_val = np.zeros(N).reshape(N,1)
    Y_train = Y_train.reshape(N,1) #Asigna espacio correctamente de Y
    
    # obtenemos la matriz con todas las distancias de las 1500 muestras nuevas
    # con respecto a todas las 3500 muestras de entrenamiento
    D = distance(X_val, X_train)
    D_argsort = np.argsort(D, axis = 1)
    
    #print("***********D_argsort")
    #print(D_argsort[0:10, : ])   
    
    Nt = len(X_val)
 
    #Matriz con la salida i correspondiente a la posición indicada en D_argsort
    Y_argsort = Yout(k, D_argsort, Y_train)

    Y_val = np.zeros(Nt).T
    

    if(tipo == 0):
        #regresion
        for x in range (Nt):
            average = np.average(Y_argsort[x])
            Y_val[x] = average
            
    elif(tipo == 1):
        #clasificacion
        
        for x in range (Nt):
            mode = stats.mode(Y_argsort[x])           
            Y_val[x] = mode[0].astype(float)
            
    Y_val.reshape(Nt,1)
    # print(Y_val, Y_val.shape)
    #¿por qué no muestra forma correcta del vector?
    
    return Y_val #Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  

def distance(X_val, X_train):
    
    L = np.dot(X_val, X_train.T)
    tem = np.sum(X_val * X_val, axis=1).reshape([X_val.shape[0],1])
    tem2 = np.sum(X_train * X_train, axis=1).reshape([1,X_train.shape[0]])
    
    D = np.dot(tem, np.ones((1,X_train.shape[0]))) + np.dot(np.ones((X_val.shape[0],1)), tem2) - 2* L
    D = np.sqrt(D)
    
    return D

#Me retorna la salida Y_i proveniente de la matriz que contiene la posición de la distancia más cortas (vecinos más cercanos)
# de las nuevas muestras
def Yout (k, D_argsort, Y_train):
    N = len(D_argsort)    
    kn = D_argsort[ : , 0: k]
    kn = kn.reshape(len(D_argsort),k)
    Y_val = np.zeros((N,k))

  #  Y_val =np.array( [[Y_train[kn[y,x]] for x in range(k)] for y in range(N)])
    for x in range (N):
        for y in range (k):
               Y_val[x][y] = Y_train[kn[x,y]]
    
    return Y_val

## Ejercicio 3

Una vez haya completado el codigo del método de KNN, ejecute varias veces el proceso de entrenamiento y evaluacion cambiando el parametro $k$, el cual es el numero de vecinos, y complete la siguiente tabla con los valores del error de clasificación obtenidos:

In [40]:
from numpy import random
from sklearn.model_selection import train_test_split
import math


def mainKNN(k_vecinos, tipoKNN = 1):
    N = np.size(X,0)
    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    for j in range(4):
        global Xtrain
        global Xtest

        Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.25)

        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)
        
        Yest = KNN(Xtrain, Ytrain, Xtest, k_vecinos, tipoKNN)#Complete con el llamado apropiado de la función que Uds diseñaron

        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
    print("\n************************* Número vecinos: ", k_vecinos)
    print('Error durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))
    
    # redondeamos a 4 cifras decimales
    mean = round(np.mean(Error),4)
    std = round(np.std(Error),4)
    return (mean,std)

mean,std = mainKNN(2)

print(str(mean),str(std))


************************* Número vecinos:  2
Error durante la prueba = 0.26973684210526316+-0.08802031684381349
0.2697 0.088


In [36]:
print("Muestras entrenamiento: ",Xtrain.shape)
print("Muestras validación",len(Xtest))

Muestras entrenamiento:  (112, 4)
Muestras validación 38


Responda:

3.1 ¿Qué metodología de validación se usa en el experimento?

La metodología de validación implementada es **validación cruzada**, es este caso $k = 4$  folds.

3.2 ¿Cuántas muestras se usan para el entrenamiento?

Hay **112 muestras** para entrenar

3.3 ¿Cuántas muestras se usan para la validación?

En general son **36 muestras** nuevas.

In [49]:
##k_vecinos = [mainKNN(x) for x in range(1,8) ]

means = np.zeros(7);
stds = np.zeros(7);

for x in range (1,8):
    mean,std = mainKNN(x);
    means[x - 1] = mean
    stds[x - 1] = std



************************* Número vecinos:  1
Error durante la prueba = 0.20394736842105263+-0.11376063464335907

************************* Número vecinos:  2
Error durante la prueba = 0.26973684210526316+-0.08802031684381349

************************* Número vecinos:  3
Error durante la prueba = 0.26973684210526316+-0.08603090020146066

************************* Número vecinos:  4
Error durante la prueba = 0.2763157894736842+-0.0842516347030638

************************* Número vecinos:  5
Error durante la prueba = 0.26315789473684215+-0.09304036594559835

************************* Número vecinos:  6
Error durante la prueba = 0.2763157894736842+-0.0842516347030638

************************* Número vecinos:  7
Error durante la prueba = 0.26973684210526316+-0.08802031684381349


#### Tabla de resultados

In [48]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types = pd.DataFrame({
    'Numero de vecinos' : pd.Series(['1', '2', '3', '4', '5', '6', '7'])})
df_types["Error_Prueba"] = means
df_types["Desviación estándar del error"] = stds
df_types.set_index(['Numero de vecinos'], inplace=True)
df_types["Error_Prueba"][1] = "0.2697"
df_types["Desviación estándar del error"][1] = "0.088"
#df_types.sort_index(inplace=True)
qgrid_widget = qgrid.show_grid(df_types, show_toolbar=False)
qgrid_widget

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': True, 'defau…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [50]:
qgrid_widget.get_changed_df()

Error_Prueba  Desviación estándar del error
Numero de vecinos                                             
1                        0.2039                         0.1138
2                        0.2697                         0.0880
3                        0.2697                         0.0860
4                        0.2763                         0.0843
5                        0.2632                         0.0930
6                        0.2763                         0.0843
7                        0.2697                         0.0880

## Ejercicio 4

A continuación debe completar el código del método parzen para resolver problemas de clasificación con el modelo ventana de Parzen.

In [114]:
#Calcula el kernel gaussiano de x
def kernel_gaussiano(x):
    return np.exp((-0.5)*x**2)

def ParzenWindow(x,Data,h):
    h = h
    Ns = Data.shape[0]
    suma = 0
    for k in range(Ns):
        u = sc.spatial.distance.euclidean(x,Data[k,:])
        suma += kernel_gaussiano(u/h)
    return suma

def ParzenClass(X_train, Y_train, X_val, ancho_h):
    #Tamaño muestras nuevas
    Nx = X_val.shape[0]
    Yest = np.zeros(Nx)
    h = ancho_h
    
    # ***** MODO GENENRAL *****
    
    # obtenemos el arreglo de clases
    Nclases = np.unique(Y_train)
    # numero de clases
    lengthClases = len(Nclases)
    # imprimimos las clases
#    print(Nclases)
    
    X_aux =[] 
#    print(X_aux.shape)
    j = 0
    for i in Nclases:
        aux = X_train[Y_train == i,:]
        X_aux.append(aux)
    
    #print(X_aux[1].shape)
    pt = np.zeros(Nx)
    for i in range(Nx):
        for j in range(len(Nclases)):
            pt[j] = ParzenWindow(X_val[i],X_aux[j], h)
        
        clase_max = np.argmax(pt)
        Yest[i] = clase_max
        
    # ********* MODO CASO DE ESTUDI= ***********
    # Separar las clases
    #clase1 = X_train[Y_train == 0]
    #clase2 = X_train[Y_train == 1]
    #clase3 = X_train[Y_train == 2]
    
    #p1= np.zeros(Nx)
    #p2= np.zeros(Nx)   
    #p3= np.zeros(Nx)
    
    #for i in range(Nx):        
    #p1[i] = ParzenWindow(X_val[i],clase1, h)
     #   p2[i] = ParzenWindow(X_val[i],clase2, h)
      #  p3[i] = ParzenWindow(X_val[i],clase3, h)        
       # p = np.array([p1[i], p2[i], p3[i]])
        #clase_max = np.argmax(p)
        #Yest[i] = clase_max 
    
    return Yest#Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  



## Ejercicio 5

Ahora debe resolver el mismo problema de clasificación con el método ventana de Parzen. Complete el código ejecute las pruebas para diferentes valores de $h$ y llene la siguiente tabla:

In [115]:
from numpy import random
from sklearn.model_selection import StratifiedKFold
import math

def mainPW(h=0.1):
    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    skf = StratifiedKFold(n_splits=4)
    j = 0
    for train, test in skf.split(X, Y):
        Xtrain = X[train,:]
        Ytrain = Y[train]
        Xtest = X[test,:]
        Ytest = Y[test]
        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        Yest = ParzenClass(Xtrain, Ytrain, Xtest,h)#Complete con el llamado apropiado de la función que Uds diseñaron
     
        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
        j += 1
    
    # redondeamos a 4 cifras decimales
    mean = round(np.mean(Error),4)
    std = round(np.std(Error),4)
    return (mean,std)

mean, std = mainPW()
print('Error durante la prueba = ' + str(mean) + '+-' + str(std))

Error durante la prueba = 0.2217+-0.0433


In [116]:
print("Muestras entrenamiento ",Xtrain.shape)
print("Muestras para validación ",Xtest.shape)

Muestras entrenamiento  (112, 4)
Muestras para validación  (38, 4)


Responda:
    
5.1 ¿Qué metodología de validación se usó en la simulación?:

**Validación cruzada estratificada** con $k = 4$ splits, valida que las particiones aleatorias contenga proporcialmente valores de cada categoría


5.2 ¿Cuántas muestras se usan para el entrenamiento?:

Se destinan *114 muestras* para entrenamiento


5.3 ¿Cuántas muestras se usan para la validación?:

Se cuenta con *36 muestras* para validación

In [117]:
ancho_h = [0.05, 0.1, 0.5, 1, 2, 5, 10 ]
# pw = [mainPW(x) for x in ancho_h]

meansPW = np.zeros(len(ancho_h));
stdsPW = np.zeros(len(ancho_h));

for x in range(0,len(ancho_h)):
    mean,std = mainPW(ancho_h[x]);
    meansPW[x - 1] = mean
    stdsPW[x - 1] = std

#### Tabla de resultados

In [118]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types2 = pd.DataFrame({
    'Tamano de ventana' : pd.Series(['0.05', '0.1', '0.5', '1', '2', '5', '10'])})
df_types2["Error_Prueba"] = meansPW
df_types2["Desviación estándar del error"] = stdsPW
df_types2.set_index(['Tamano de ventana'], inplace=True)
df_types2["Error_Prueba"][1] = "0.222"
df_types2["Desviación estándar del error"][1] = "0.043"
#df_types.sort_index(inplace=True)
qgrid_widget = qgrid.show_grid(df_types2, show_toolbar=False)
qgrid_widget

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': True, 'defau…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [119]:
qgrid_widget.get_changed_df()

Error_Prueba  Desviación estándar del error
Tamano de ventana                                             
0.05                     0.2217                         0.0433
0.1                      0.2220                         0.0430
0.5                      0.2810                         0.0315
1                        0.2420                         0.0585
2                        0.2217                         0.0470
5                        0.2217                         0.0470
10                       0.2217                         0.0433